### PR TITLE
added implementation for store data to file system

### PR DESCRIPTION
**Pull request by akorneta**

### CODE REVIEW

File: platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/LocalPreferenceDaoImpl.java [hunk 1]
```diff
@@ -43,92 +35,40 @@
...
+    private void stop() throws IOException {
+        preferenceStorage.store(preferences);
     }
 
     @Override
     public void setPreferences(String userId, Map<String, String> preferences) throws ServerException, NotFoundException {
         lock.writeLock().lock();
         try {
-            storage.put(userId, new HashMap<>(preferences));
+            this.preferences.put(userId, new HashMap<>(preferences));
```
Comment (garagatyi) [inline platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/LocalPreferenceDaoImpl.java]: it is too easy to make mistake here. What if rename one of the preferences variables ?

File: platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/LocalUserDaoImpl.java [hunk 2]
```diff
@@ -28,13 +35,16 @@
...
+    private final LocalStorage  userStorage;
 
     @Inject
-    public LocalUserDaoImpl(@Named("codenvy.local.infrastructure.users") Set<User> users) {
+    public LocalUserDaoImpl(@Named("codenvy.local.infrastructure.users") Set<User> users, LocalStorageFactory storageFactory) {
         this.users = new LinkedList<>();
         lock = new ReentrantReadWriteLock();
+        userStorage = storageFactory.create("users.json");
         try {
             for (User user : users) {
```
Comment (garagatyi) [inline platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/LocalUserDaoImpl.java]: Why we have here user storage and list of users at the same time?

File: platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/LocalWorkspaceDaoImpl.java [hunk 3]
```diff
@@ -28,15 +35,18 @@
...
+    private final LocalStorage    workspaceStorage;
 
     @Inject
-    public LocalWorkspaceDaoImpl(@Named("codenvy.local.infrastructure.workspaces") Set<Workspace> workspaces) {
+    public LocalWorkspaceDaoImpl(@Named("codenvy.local.infrastructure.workspaces") Set<Workspace> workspaces, LocalStorageFactory storageFactory) {
         this.workspaces = new LinkedList<>();
         lock = new ReentrantReadWriteLock();
+        workspaceStorage = storageFactory.create("workspaces.json");
         try {
             for (Workspace workspace : workspaces) {
```
Comment (garagatyi) [inline platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/LocalWorkspaceDaoImpl.java]: Why we have here workspace storage and list of workspaces at the same time?

File: platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/storage/LocalStorage.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+package org.eclipse.che.api.local.storage;
+
+import com.google.gson.reflect.TypeToken;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Provides methods for store and retrieve objects from local storage.
```
Comment (garagatyi) [inline platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/storage/LocalStorage.java]: methods for storing
or
methods to store

File: platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/storage/LocalStorageImpl.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+public class LocalStorageImpl implements LocalStorage {
+
+    private static final Logger LOG  = LoggerFactory.getLogger(LocalStorageImpl.class);
+    private static final Gson   GSON = new Gson();
+
+    /** json file to store and load */
+    private final File storedFile;
+
+    @Inject
+    public LocalStorageImpl(@Named("user.local.db") String rootDirPath, @Assisted String fileName)
```
Comment (garagatyi) [inline platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/storage/LocalStorageImpl.java]: property name doesn't say that its value is path to dir

File: platform-api/che-core-api-infrastructure-local/pom.xml [hunk 6]
```diff
@@ -22,6 +22,16 @@
     <name>Che Core :: API :: Local Infrastructure</name>
     <dependencies>
         <dependency>
+            <groupId>com.google.inject</groupId>
+            <artifactId>guice</artifactId>
```
Comment (skabashnyuk) [inline platform-api/che-core-api-infrastructure-local/pom.xml]: why do you need it?

File: platform-api/che-core-api-infrastructure-local/pom.xml [hunk 7]
```diff
@@ -22,6 +22,16 @@
...
+            <version>${com.google.code.guice.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>com.google.inject.extensions</groupId>
+            <artifactId>guice-assistedinject</artifactId>
+            <version>${com.google.code.guice.version}</version>
```
Comment (skabashnyuk) [inline platform-api/che-core-api-infrastructure-local/pom.xml]: why do you need it?

File: platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/InMemoryFactoryStore.java [hunk 8]
```diff
@@ -36,9 +43,33 @@
  */
 @Singleton
 public class InMemoryFactoryStore implements FactoryStore {
```
Comment (evoevodin) [inline platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/InMemoryFactoryStore.java]: I think it should be renamed to <code>LocalFactoryStore</code> as it stores data to fs

File: platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/storage/LocalStorageFactory.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+ * @author Anton Korneta
+ */
+@Singleton
+public class LocalStorageFactory {
+
+    @Inject
+    @Named("local.storage.path")
+    private String pathToStorage;
+
+    public LocalStorage create(String fileName) throws IOException {
```
Comment (evoevodin) [inline platform-api/che-core-api-infrastructure-local/src/main/java/org/eclipse/che/api/local/storage/LocalStorageFactory.java]: add javadoc

**Pull request merged**